platform/pwl.c

```diff
@@ -39,3 +61,775 @@ pwl_display_destroy (PwlDisplay *self)
+        clear_egl ();
+        return FALSE;
+    }
+
+    return TRUE;
+}
+
+void clear_egl (void)
+{
+    if (egl_data.display != EGL_NO_DISPLAY) {
+        if (egl_data.context != EGL_NO_CONTEXT) {
+            eglDestroyContext (egl_data.display, egl_data.context);
+        }
+        eglTerminate (egl_data.display);
+        egl_data.context = EGL_NO_CONTEXT;
+        egl_data.display = EGL_NO_DISPLAY;
+    }
+    eglReleaseThread ();
+}
+
+
+gboolean init_wayland (GError **error)
+{
+    g_debug ("Initializing Wayland...");
+
+    if (!(wl_data.display = wl_display_connect (NULL))) {
+        g_set_error (error,
+                     G_FILE_ERROR,
+                     g_file_error_from_errno (errno),
+                     "Could not open Wayland display");
+        return FALSE;
+    }
+
+    wl_data.registry = wl_display_get_registry (wl_data.display);
+    g_assert (wl_data.registry);
+    wl_registry_add_listener (wl_data.registry,
+                              &registry_listener,
+                              NULL);
+    wl_display_roundtrip (wl_data.display);
+
+    g_assert (wl_data.compositor);
+    g_assert (wl_data.xdg_shell != NULL ||
+              wl_data.shell != NULL ||
+              wl_data.fshell != NULL);
+
+    return TRUE;
+}
+
+void clear_wayland (void)
+{
+    g_source_destroy (wl_data.event_src);
+
+    if (wl_data.xdg_shell != NULL)
+        xdg_wm_base_destroy (wl_data.xdg_shell);
+    if (wl_data.fshell != NULL)
+        zwp_fullscreen_shell_v1_destroy (wl_data.fshell);
+    if (wl_data.shell != NULL)
+        wl_shell_destroy (wl_data.shell);
+
+    if (wl_data.compositor != NULL)
+        wl_compositor_destroy (wl_data.compositor);
+
+    wl_registry_destroy (wl_data.registry);
+    wl_display_flush (wl_data.display);
+    wl_display_disconnect (wl_data.display);
+}
+
+static void
+xdg_surface_on_configure (void *data,
+                          struct xdg_surface *surface,
+                          uint32_t serial)
+{
+    xdg_surface_ack_configure (surface, serial);
+}
+
+static const struct xdg_surface_listener xdg_surface_listener = {
+    .configure = xdg_surface_on_configure
+};
+
+static void
+xdg_toplevel_on_configure (void *data,
+                           struct xdg_toplevel *toplevel,
+                           int32_t width, int32_t height,
+                           struct wl_array *states)
+{
+    configure_surface_geometry (width, height);
+
+    g_debug ("New XDG toplevel configuration: (%" PRIu32 ", %" PRIu32 ")", width, height);
+
+    wl_data.resize_window(data);
+}
+
+static void
+xdg_toplevel_on_close (void *data, struct xdg_toplevel *xdg_toplevel)
+{
+    // TODO g_application_quit (g_application_get_default ());
+}
+
+static const struct xdg_toplevel_listener xdg_toplevel_listener = {
+    .configure = xdg_toplevel_on_configure,
+    .close = xdg_toplevel_on_close,
+};
+
+
+gboolean create_window (void *data, GError **error)
```

Comment:
As you can imagine, most of this should be moved into a `pwl_window_create()` function :wink: